### PR TITLE
refactor: ProjectConfig and its usage

### PR DESCRIPTION
**Pull request by o-andrieiev**
Introduce new type for ProjectConfig
This does not introduce any runtime overhead, and is mostly needed for static typing
Text description from 'configuring' should be moved here eventually

### CODE REVIEW

File: universum/configuration_support.py [hunk 1]
```diff
@@ -13,6 +16,47 @@
 skip_attributes = {"children", "critical", "skip_numbering_level"}
 
 
+ProjectConfigKeyType = Literal["name", "command", "environment", "artifacts", "report_artifacts",
+                               "artifact_prebuild_clean", "directory", "critical", "background", "finish_background",
+                               "code_report", "pass_tag", "fail_tag", "if_env_set"]
+
+
+class ProjectConfig(TypedDict, total=True):
```
Comment (i-keliukh) [inline universum/configuration_support.py:24]: Why total=True?

File: universum/configuration_support.py [hunk 2]
```diff
@@ -13,6 +16,47 @@
...
+    """
+    :param name: Specifies the name of the configuration.
+    :type name: str
+    The name is used as the title of the build log block corresponding to the build of this configuration. It is
+    also used to generate name of the log file if the option for storing to log files is selected. A project
+    configuration can have no name, but it is not recommended for aesthetic reasons. If several project
+    configurations have the same names, and logs are stored to files (see --launcher-output / -lo command-line
```
Comment (k-dovgan) [inline universum/configuration_support.py:31]: ``--launcher-output`` / ``-lo`` still looks ok in docstrings, and might look way better for generated documentation

File: universum/configuration_support.py [hunk 3]
```diff
@@ -13,6 +16,47 @@
...
+    parameter for details), all logs for such configurations will be stored to one file in order of their appearances.
+
+    :param command: The list with command line for the configuration launch.
+    :type command: list
```
Comment (k-dovgan) [inline universum/configuration_support.py:35]: Shouldn't it be, like, 'list of str', or something alike? It must not be just a list of anything...

File: universum/configuration_support.py [hunk 4]
```diff
@@ -13,6 +16,47 @@
...
+    For every project configuration first list item should be a console command (e.g. script name, or any other
+    command like ls), and other list items should be the arguments, added to the command (e.g. --debug or -la). Every
+    command line element, separated with space character, should be passed as a separate string argument. Lists like
+    ["ls -a"] will not be processed correctly and thus should be splat into ["ls", "-a"]. Lists like ["build.sh",
+    "--platform A"] will not be processed correctly and thus should be plat into ["build.sh", "--platform",
+    "A"]. A project configuration can have an empty list as a command. Such configuration won’t do anything except
+    informing user about missing it.
```
Comment (k-dovgan) [inline universum/configuration_support.py:42]: Looks like you took the description from documentation; but it doesn't look nice now that I'm reading it. Can you, maybe, rephrase it a little to make it more understandable? Especially the thing about spaces and _arguments_
(You may even apply the same changes to actual documentation paragraph you took this description from in the same PR)

File: universum/configuration_support.py [hunk 5]
```diff
@@ -13,6 +16,47 @@
...
+    """
```
Comment (k-dovgan) [inline universum/configuration_support.py:43]: Will further descriptions be added here later?

File: universum/configuration_support.py (hunk 1 above)
Comment (o-andrieiev) [reply universum/configuration_support.py:24]: fixed.

File: universum/configuration_support.py (hunk 2 above)
Comment (o-andrieiev) [reply universum/configuration_support.py:31]: I don't follow this comment. Would you like me to change something?

File: universum/configuration_support.py (hunk 4 above)
Comment (o-andrieiev) [reply universum/configuration_support.py:42]: ok, I'll include these changes in the PR

File: universum/configuration_support.py (hunk 5 above)
Comment (o-andrieiev) [reply universum/configuration_support.py:43]: yep, I just submitted first take, since I wasn't sure whether I should continue with this work.
Having ProjectConfigKeyType and ProjectConfig keys as separate items is not ideal.
However this change does let us move all the information from configuring.html to configuration_support.html, removing the former page completely.

File: universum/configuration_support.py (hunk 2 above)
Comment (k-dovgan) [reply universum/configuration_support.py:31]: Yes, I meant changing '-quotes to `-quotes exactly for reasons demonstrated in comment above.

Comment (o-andrieiev) [reply universum/configuration_support.py]: I just fully copied descriptions from 'configuring.rst' instead

File: universum/configuration_support.py (hunk 5 above)
Comment (o-andrieiev) [reply universum/configuration_support.py:199]: yep, I just submitted first take, since I wasn't sure whether I should continue with this work. Having ProjectConfigKeyType and ProjectConfig keys as separate items is not ideal. However this change does let us move all the information from configuring.html to configuration_support.html, removing the former page completely.

**Comment by k-dovgan** [comment]
https://universum.readthedocs.io/en/type-annot/

File: universum/configuration_support.py [hunk 6]
```diff
@@ -12,6 +15,203 @@
 
 skip_attributes = {"children", "critical", "skip_numbering_level"}
 
+ProjectConfigKeyType = Literal["name", "command", "environment", "artifacts", "report_artifacts",
+                               "artifact_prebuild_clean", "directory", "critical", "background", "finish_background",
+                               "code_report", "pass_tag", "fail_tag", "if_env_set"]
+
+
+class ProjectConfig(TypedDict, total=False):
```
Comment (i-keliukh) [inline universum/configuration_support.py:23]: 1. Since users are not going to use this class, it has to be removed from the documentation.
Alternatively, we can change the approach to creating configs and let users to use this class.
However, this would be breaking change.
2. If we are discussing breaking changes, maybe it would be better to move away from a dictionary.
As an option, Variations can convert dictionary to the class automatically, but accept both variants.

File: universum/configuration_support.py [hunk 7]
```diff
@@ -12,6 +15,203 @@
...
+    report_artifacts: List[str]
+    artifact_prebuild_clean: bool
+    directory: str
+    critical: bool
+    background: bool
+    finish_background: bool
+    code_report: str
+    pass_tag: str
+    fail_tag: str
+    if_env_set: str
```
Comment (i-keliukh) [inline universum/configuration_support.py:213]: 1. Are you sure we are not limiting users to the listed key names?
Currently we are ignoring all unknown key names and, as fat as I know, some users already are using custom keys.
2. Can we avoid duplication between this list and ProjectConfigKeyType?

File: universum/configuration_support.py [hunk 8]
```diff
@@ -12,6 +15,203 @@
 
 skip_attributes = {"children", "critical", "skip_numbering_level"}
 
+ProjectConfigKeyType = Literal["name", "command", "environment", "artifacts", "report_artifacts",
```
Comment (i-keliukh) [inline universum/configuration_support.py:18]: The benefits of using this annotation are not clear to me.
There is only one usage in the code with only one key name checked.
My guess is that you are going to expand the usage in further updates.
Please confirm it.

File: universum/configuration_support.py (hunk 6 above)
Comment (o-andrieiev) [reply universum/configuration_support.py]: as temporary solution, I kept inheritance from dict, but I intend to remove that dependency in this PR, after I rewrite all other modules to use properties instead of accessing dict keys. I already updated Variations to accept either dict or ProjectConfiguration - which is the only change that would require new test case I think.

File: universum/configuration_support.py (hunk 5 above)
Comment (o-andrieiev) [reply universum/configuration_support.py:194]: I no longer think that we can remove configuring.html completely, but the "common keys" will move to configuration_support.html as part of ProjectConfiguration class description

File: tests/deployment_utils.py [hunk 9]
```diff
@@ -128,7 +128,7 @@ def exit(self):
 
 
 @pytest.fixture()
-def execution_environment(request) -> ExecutionEnvironment:
+def execution_environment(request) -> Generator[ExecutionEnvironment, None, None]:
```
Comment (i-keliukh) [inline tests/deployment_utils.py:131]: 1. This is unrelated change. In general I am ok to mix annotation in different modules inside one commit, but configs change to class is too big change on its own.
2. pytest fixture are quite special beasts. I don't think we could benefit from this annotation anyway. Does mypy know that the function here and the parameter in test function are connected?

File: universum/configuration_support.py [hunk 10]
```diff
@@ -317,57 +422,73 @@ class Variations(list):
 
     """
 
-    def __add__(self, other: List) -> 'Variations':
+    def __init__(self, lst: Optional[Union[List[Dict[str, Any]], List[ProjectConfiguration]]]):
+        #  lst can be List[Dict[str, Any]] to support legacy cases - should be removed after project migration
+        self.configs: List[ProjectConfiguration] = []  # aggregation is used instead of inheritance for type safety
```
Comment (i-keliukh) [inline universum/configuration_support.py:427]: Why can't we just inherit from list?

File: tests/deployment_utils.py (hunk 9 above)
Comment (o-andrieiev) [reply tests/deployment_utils.py:131]: mypy flagged these 2 lines as errors, when I checked the test files. I am fine with moving this change to separate PR.

File: universum/configuration_support.py (hunk 10 above)
Comment (i-keliukh) [reply universum/configuration_support.py:426]: Answer: there is no way to inherit from typed list

File: universum/configuration_support.py [hunk 11]
```diff
@@ -117,57 +421,73 @@ class Variations(list):
...
         :return: iterable for all dictionary objects in :class:`.Variations` list
         """
-        for obj_a in self:
+        for obj_a in self.configs:
 
             obj_a_copy = copy.deepcopy(obj_a)
 
-            if "children" in obj_a_copy:
-                for obj_b in obj_a_copy["children"].all():
+            if obj_a_copy.children:
```
Comment (i-keliukh) [inline universum/configuration_support.py:489]: Is it equivalent change?

File: universum/modules/code_report_collector.py [hunk 12]
```diff
@@ -39,29 +32,18 @@ def set_code_report_directory(self, project_root: str) -> None:
         if not os.path.exists(self.report_path):
             os.makedirs(self.report_path)
 
-    def prepare_environment(self, project_configs: List[ProjectConfig]) -> Variations:
-        afterall_steps: List[ProjectConfig] = []
-        for item in project_configs:
-            if not item.get("code_report", False):
+    def prepare_environment(self, project_config_variations: Variations) -> Variations:
+        afterall_steps: List[ProjectConfiguration] = []
```
Comment (i-keliukh) [inline universum/modules/code_report_collector.py:36]: Would it be safe to use Variations here?

File: universum/modules/launcher.py [hunk 13]
```diff
@@ -350,30 +357,27 @@ def __init__(self, *args, **kwargs):
...
+        self.code_report_collector = self.code_report_collector_factory()
         self.include_patterns, self.exclude_patterns = get_match_patterns(self.settings.step_filter)
 
     @make_block("Processing project configs")
-    def process_project_configs(self):
+    def process_project_configs(self) -> configuration_support.Variations:
         config_path = utils.parse_path(self.settings.config_path, self.settings.project_root)
         configuration_support.set_project_root(self.settings.project_root)
-        config_globals = {}
+        config_globals: Dict[str, configuration_support.Variations] = {}
```
Comment (i-keliukh) [inline universum/modules/launcher.py:367]: I am not sure regarding technical consequences of this declaration, but config_globals is just a dictionary of all global variables of the configs module. Potentially it can contain variable of any type.

File: universum/configuration_support.py (hunk 11 above)
Comment (o-andrieiev) [reply universum/configuration_support.py:489]: it's not equivalent, but I think this change is warranted: previously configuration could lack some fields, and universum had lots of 'get' calls with provided default values - this was risky in case default was incorrectly set to 2 different values in 2 different parts of code. Instead ProjectConfiguration is constructed with all properties set, but all of them are 'falsy' by default. This means that there's an internal guarantee of returned type (no 'Optional'), which simplifies the code.

File: universum/modules/code_report_collector.py (hunk 12 above)
Comment (o-andrieiev) [reply universum/modules/code_report_collector.py:36]: in afterall_steps ?

File: universum/modules/launcher.py (hunk 13 above)
Comment (o-andrieiev) [reply universum/modules/launcher.py:367]: sure, I can remove this part, and add explicit type cast, where Variations are retrieved from dict

File: universum/modules/code_report_collector.py (hunk 12 above)
Comment (i-keliukh) [reply universum/modules/code_report_collector.py:36]: yes

File: universum/configuration_support.py (hunk 11 above)
Comment (i-keliukh) [reply universum/configuration_support.py:489]: Ok, just to make it clear: are you saying that obj_a_copy is guaranteed to have "children" field?
If it is guaranteed by the static typing, I'm ok with the change.

File: universum/modules/code_report_collector.py (hunk 12 above)
Comment (o-andrieiev) [reply universum/modules/code_report_collector.py:36]: I am not sure I understand, we can either add Variations, or populate list of ProjectConfigurations and convert it at the end - there should be no difference really. I think current approach (convert in the end) has slightly lower overhead, since adding Variations has one more call to go through.

File: universum/configuration_support.py (hunk 11 above)
Comment (o-andrieiev) [reply universum/configuration_support.py:489]: children is a property so it's guaranteed, but it can return None which is falsy

File: universum/modules/code_report_collector.py (hunk 12 above)
Comment (i-keliukh) [reply universum/modules/code_report_collector.py:36]: It's about consistency (readability). I interpret Variations as a type used to store list of ProjectConfiguration instances, and the performance difference is minuscule.
BTW, probably we should also rename or alias Variations to something more meaningful.

Comment (o-andrieiev) [reply universum/modules/code_report_collector.py:36]: sure, I'll change it to Variations. I think there were several other places where that would make sense.
I propose following naming options: ConfigurationList, ConfigurationHolder.

Comment (i-keliukh) [reply universum/modules/code_report_collector.py:36]: ConfigurationList looks good to me

File: universum/modules/launcher.py [hunk 14]
```diff
@@ -22,14 +23,14 @@
...
+def make_command(name: str) -> sh.Command:
     try:
         return sh.Command(name)
-    except sh.CommandNotFound as e :
+    except sh.CommandNotFound as e:
         raise CiException(f"No such file or command as '{name}'") from e
 
 
-def check_if_env_set(configuration):
+def check_if_env_set(configuration: configuration_support.ProjectConfiguration) -> bool:
```
Comment (o-andrieiev) [inline universum/modules/launcher.py:33]: move to ProjectConfiguration class?

File: universum/configuration_support.py [hunk 15]
```diff
@@ -1,19 +1,286 @@
...
+                result = True
+            command_line = command_line + " " if command_line else "" + argument
+        self.command = [command_line]
+        return result
+
+
+DictType = TypeVar('DictType', bound=dict)
+
+
+def combine(dictionary_a: DictType, dictionary_b: DictType) -> DictType:
```
Comment (o-andrieiev) [inline universum/configuration_support.py:282]: mark deprecated?

File: universum/configuration_support.py [hunk 16]
```diff
@@ -1,19 +1,286 @@
...
+
+    def stringify_command(self) -> bool:
+        result = False
+        command_line = ""
+        for argument in self.command:
+            if " " in argument:
+                argument = "\"" + argument + "\""
+                result = True
+            command_line = command_line + " " if command_line else "" + argument
+        self.command = [command_line]
```
Comment (o-andrieiev) [inline universum/configuration_support.py:275]: self.comman is List of strings, but this used to assign one string. Change to Union or report as list of 1 element?

File: universum/configuration_support.py [hunk 17]
```diff
@@ -1,19 +1,286 @@
...
+        self.extras: Dict[str, str] = extras if extras else {}
+        for key, value in kwargs.items():
+            self.extras[key] = value
+
+    def replace_string(self, from_string: str, to_string: str) -> None:
+        self.command = [word.replace(from_string, to_string) for word in self.command]
+        self.artifacts = self.artifacts.replace(from_string, to_string)
+        self.report_artifacts = self.report_artifacts.replace(from_string, to_string)
+        self.directory = self.directory.replace(from_string, to_string)
+        # should we update extras? (user-defined)
```
Comment (o-andrieiev) [inline universum/configuration_support.py:236]: should we replace path in user-defined variables?

File: universum/configuration_support.py (hunk 4 above)
Comment (o-andrieiev) [reply universum/configuration_support.py]: it was decided to move documentation changes in separate PR

File: universum/configuration_support.py (hunk 17 above)
Comment (k-dovgan) [reply universum/configuration_support.py:63]: I'd say 'yes'. As a user, I'd expect a 'variable-like' to behave as a variable always, not only in some cases.

File: universum/modules/launcher.py (hunk 14 above)
Comment (k-dovgan) [reply universum/modules/launcher.py:33]: Maybe, but not in this PR, that's a separate change.

File: doc/configuring.rst [hunk 18]
```diff
@@ -311,11 +311,12 @@ pass_tag, fail_tag
...
     a tag with spaces in TeamCity's web-interface. Every tag is added (if matching condition) after executing
     build step it is set in, not in the end of all run.
-    ``pass_tag`` and ``fail_tag`` can also be used when `multiplying build configurations`_, like this:
+    ``pass_tag`` and ``fail_tag`` can also be used when
+    `multiplying build configuration <configuring.html#multiplying-build-configurations>`__, like this:
 
     .. testsetup::
 
-        #!/usr/bin/env {python}
+        #!/usr/bin/env python3.7
```
Comment (o-andrieiev) [inline doc/configuring.rst:319]: this was unintentional

File: universum/configuration_support.py [hunk 19]
```diff
@@ -1,18 +1,120 @@
...
+
+    def __repr__(self) -> str:
+        res = {k: v for k, v in self.__dict__.items() if v and k != 'extras'}
+        res.update(self.extras)
+        if len(self.command) == 1:  # command should be printed as one string, instead of list
```
Comment (o-andrieiev) [inline universum/configuration_support.py:68]: this way we don't have to use Union[str, List[str]] for command attribute, but can still print ProjectConfiguration cleanly

File: universum/configuration_support.py [hunk 20]
```diff
@@ -1,18 +1,120 @@
...
+            res['command'] = self.command[0]
+        return str(res)
+
+    def __eq__(self, other: Any) -> bool:
```
Comment (o-andrieiev) [inline universum/configuration_support.py:72]: this is to preserve ability to compare Variations with list[dict] (old format)

File: universum/configuration_support.py [hunk 21]
```diff
@@ -1,18 +1,120 @@
...
+        if isinstance(other, dict):
+            for key, val in other.items():
+                if getattr(self, key, None) != val:
+                    return False
+            return True
+        return super().__eq__(other)
+
+    def __getitem__(self, item: str) -> Optional[str]:
```
Comment (o-andrieiev) [inline universum/configuration_support.py:80]: this is to simulate dict-like access for other clients

File: universum/configuration_support.py [hunk 22]
```diff
@@ -1,18 +1,120 @@
...
+            background=self.background or other.background,
+            finish_background=self.finish_background or other.finish_background,
+            code_report=self.code_report or other.code_report,
+            pass_tag=self.pass_tag + other.pass_tag,
+            fail_tag=self.fail_tag + other.fail_tag,
+            if_env_set=self.if_env_set + other.if_env_set,
+            extras=combine(self.extras, other.extras)
+        )
+
+    def stringify_command(self) -> bool:
```
Comment (o-andrieiev) [inline universum/configuration_support.py:101]: this is such a weird function honestly

File: universum/modules/code_report_collector.py [hunk 23]
```diff
@@ -39,30 +32,19 @@ def set_code_report_directory(self, project_root: str) -> None:
...
+    def prepare_environment(self, project_config_variations: Variations) -> Variations:
+        afterall_steps: Variations = Variations()
+        for item in project_config_variations.configs:
+            if not item.code_report:
                 continue
 
-            self.set_code_report_directory(self.settings.project_root)
-            temp_filename: str = "${CODE_REPORT_FILE}"
-            name: str = utils.calculate_file_absolute_path(self.report_path, item.get("name", "")) + ".json"
+            self.set_code_report_directory(self.settings.project_root)  # why is this in a loop?
```
Comment (o-andrieiev) [inline universum/modules/code_report_collector.py:41]: why is this in a loop?

File: tests/deployment_utils.py [hunk 24]
```diff
@@ -4,11 +4,11 @@
 import os
 from pwd import getpwnam
 
+from typing import Generator
 import docker
 import py
 import pytest
 from requests.exceptions import ReadTimeout
-
```
Comment (k-dovgan) [inline tests/deployment_utils.py:11]: Why did you remove this line? It was separating absolute imports from related, that's what we do, like, everywhere in this project.

File: universum/configuration_support.py (hunk 17 above)
Comment (o-andrieiev) [reply universum/configuration_support.py]: fixed in altest update

File: tests/deployment_utils.py (hunk 24 above)
Comment (o-andrieiev) [reply tests/deployment_utils.py]: by mistake, fixed now

File: universum/modules/vcs/perforce_vcs.py [hunk 25]
```diff
@@ -510,7 +510,7 @@ def map_local_path_to_depot(self, report):
                 abs_path = self.p4.run("where", line["depotFile"])[0]["path"]
                 self.mappings_dict[abs_path] = line["depotFile"]
 
-    @make_block("Revert workspace to depot state")
+    @make_block("Revert workspace to depot state", pass_errors=False)
```
Comment (o-andrieiev) [inline universum/modules/vcs/perforce_vcs.py:513]: actually, I'm not sure why 'pass_errors=False' was missing here before

**Review by k-dovgan — DISMISSED** [review]

**Review by miltolstoy — DISMISSED** [review]

File: doc/configuring.rst [hunk 26]
```diff
@@ -224,7 +224,7 @@ artifacts
     Path to the file or directory to be copied to the working directory as an execution result.
     Can contain shell-style pattern matching (e.g. `"out/*.html"`), including recursive wildcards
     (e.g. `"out/**/index.html"`).
-    If not stated otherwise (see ``--no-archive`` `command-line parameter <args.html#Artifact\ collection>`__
+    If not stated otherwise (see ``--no-archive`` `command-line parameter <args.html#Artifact\\ collection>`__
```
Comment (i-keliukh) [inline doc/configuring.rst:227]: The added slash break the link, see it here: https://universum.readthedocs.io/en/type-annot/configuring.html

**Review by k-dovgan — APPROVED** [review]

File: universum/modules/code_report_collector.py (hunk 23 above)
Comment (i-keliukh) [reply universum/modules/code_report_collector.py:41]: my guess it is "legacy reasons"
probably in original version there were no loop

Comment (i-keliukh) [reply universum/modules/code_report_collector.py:41]: anyway, you can move it out of the loop

Comment (i-keliukh) [reply universum/modules/code_report_collector.py:41]: actually, you should move it to after the loop: as I see now, the idea was not to create directory if there are no "code report" steps

File: universum/modules/code_report_collector.py [hunk 27]
```diff
@@ -39,30 +32,19 @@ def set_code_report_directory(self, project_root: str) -> None:
...
+            temp_filename: str = "${CODE_REPORT_FILE}"  # TODO: move to global constant and update docs to use it
```
Comment (i-keliukh) [inline universum/modules/code_report_collector.py:42]: What would be the benefit of using global constant?

File: universum/configuration_support.py [hunk 28]
```diff
@@ -1,18 +1,210 @@
...
+        self.finish_background: bool = finish_background
+        self.pass_tag: str = pass_tag
+        self.fail_tag: str = fail_tag
+        self.if_env_set: str = if_env_set
+        self.children: Optional['Variations'] = None
+        self.extras: Dict[str, str] = extras if extras else {}
+        for key, value in kwargs.items():
+            self.extras[key] = value
+
+    def __repr__(self) -> str:
```
Comment (i-keliukh) [inline universum/configuration_support.py:57]: Why we need custom `__repr__` function at all?
Isn't the default implementation for class enough?

Comment (o-andrieiev) [reply universum/configuration_support.py:57]: default __repr__ returns "ProjectConfiguration object..."
Custom __repr__ helps with readable debug output + this way we can keep docstrings unchanged.

File: universum/configuration_support.py [hunk 29]
```diff
@@ -1,18 +1,210 @@
...
+        :return: client-defined value
+
+        >>> cfg = ProjectConfiguration(name='foo', my_var='bar')
+        >>> cfg['name']
+        'foo'
+        >>> cfg['my_var']
+        'bar'
+        >>> cfg['test']
+        """
+        return self.extras.get(item, self.__dict__.get(item, None))
```
Comment (i-keliukh) [inline universum/configuration_support.py:120]: From this implementation it looks like `self.extras` is kind of override for `self.__dict__`.
As can be seen in `self.__init__` there are two ways to add items into "self.extras": from explicitly passed "extras" argument and from keyword dictionary (kwargs). The first way allows to actually override "native" keys.
I my opinion there are several issues here:
 - do we need such overriding at all?
 - are we using "extras" argument anywhere?
 - "extras" work only for dict-like access (cfg['my_var']), but they don't work for object-like access (cfg.my_var)

File: universum/configuration_support.py (hunk 21 above)
Comment (i-keliukh) [reply universum/configuration_support.py:106]: What about other dict methods, such as `__len__`, `__setitem__`, `__delitem__`?
If we don't need them, we should mention this decision in the code.

File: universum/configuration_support.py (hunk 28 above)
Comment (i-keliukh) [reply universum/configuration_support.py:57]: Ok, I've got the debug output argument.
But what is the issue with docstrings exactly?

Comment (o-andrieiev) [reply universum/configuration_support.py:57]: in docstrings there are many cases where indexed Variations object is printed. Since indexing Variations should return ProjectConfiguration - we need to redefine repr so that comparisong with dict contents works in docstring tests

File: universum/configuration_support.py (hunk 21 above)
Comment (o-andrieiev) [reply universum/configuration_support.py:106]: I assume ProjectConfigurations should be read-only. I'll define __setitem__ and __delitem__ to raise NotImplementedError

File: universum/configuration_support.py (hunk 29 above)
Comment (o-andrieiev) [reply universum/configuration_support.py:120]: "extras" exists to support client-defined dict values. With extras as it is - clients should notice no changes when this patch is applied. Internally universum should never use extras.
You're right that case when "extras" is a dict key deserves special attention - I'll add more tests for that case.

Comment (o-andrieiev) [reply universum/configuration_support.py:120]: I don't think "extras" need to support object-like access, since they only exist to simulate legacy behaviour. I think best way would be to generate a warning anytime extras is used.

File: universum/configuration_support.py (hunk 28 above)
Comment (i-keliukh) [reply universum/configuration_support.py:57]: Ok, as I understand, previously it printed an array of dicts with their exact content. This is what the default `__repr__` for dict does. And since we are printing list, its `__str__` method uses `__repr__` of its elements for printing, not `__str__`. Unfortunately, the default `__repr__` for class is useless.

Please expand the explanation in the docstring.

Comment (o-andrieiev) [reply universum/configuration_support.py:57]: exactly! I'll update the description

File: universum/configuration_support.py [hunk 30]
```diff
@@ -1,18 +1,210 @@
...
+
+    def __add__(self, other: 'ProjectConfiguration') -> 'ProjectConfiguration':
+        """
+        This functions defines operator ``+`` for :class:`.ProjectConfiguration` class objects by
+        concatenating strings and contents of dictionaries
+
+        :param other: `ProjectConfiguration` object
+        :return: new `ProjectConfiguration` object, including all attributes from both `self` and `other` objects
```
Comment (i-keliukh) [inline universum/configuration_support.py:128]: Please mention that "critical" attribute is skipped.

File: universum/configuration_support.py [hunk 31]
```diff
@@ -117,7 +132,38 @@ def __getitem__(self, item: str) -> Optional[str]:
...
+        [<warnings.WarningMessage object at ...>]
+        >>> set_cfg_and_get_warnings(cfg, '_extras', 'baz')
+        []
+        >>> set_cfg_and_get_warnings(cfg, 'test', 42)
+        []
+        >>> cfg
+        {'name': 'bar', 'directory': 'foo', 'my_var': 'bar', '_extras': 'baz', 'test': 42}
+        """
+        if key in self.__dict__ and key != '_extras':
+            warn("Re-defining the value of a read-only ProjectConfiguration field", UserWarning, 3)
```
Comment (i-keliukh) [inline universum/configuration_support.py:163]: Well, the field is not read-only, it just has to be accessed using object-like access.
I would give an example of proper usage in the warning, like the following:
"Re-defining the value of ProjectConfiguration field. Please use var.command to set it instead of using var['command']."

File: universum/configuration_support.py [hunk 32]
```diff
@@ -33,8 +34,19 @@ def __init__(self,
...
                  fail_tag: str = '',
                  if_env_set: str = '',
-                 extras: Optional[Dict[str, str]] = None,
-                 **kwargs) -> None:  # explicit constructor for client's type safety
+                 **kwargs) -> None:
+        """
+        All params supplied via named attributed shall be  type-checked for safety. Clients may provide custom
+        parameters in kwargs - these parameters will be stored internally in '_extras' dict and can be retrieved by
+        indexing.
+        >>> cfg = ProjectConfiguration(name='foo', command=['1', '2', '3'], _extras={'_extras': 'test', 'bool': True})
```
Comment (i-keliukh) [inline universum/configuration_support.py:42]: The doctest is confusing because you used the same name for the dictionary to hold extra values and for the extra value itself.

File: universum/configuration_support.py (hunk 21 above)
Comment (o-andrieiev) [reply universum/configuration_support.py:121]: defined setitem. Should delitem be defined still?

File: universum/configuration_support.py (hunk 31 above)
Comment (o-andrieiev) [reply universum/configuration_support.py:168]: fixed

File: universum/configuration_support.py (hunk 32 above)
Comment (o-andrieiev) [reply universum/configuration_support.py:42]: this was intentional to show that nothing breaks internally if names match

Comment (o-andrieiev) [reply universum/configuration_support.py:42]: e.g. the clients/user doesn't need to care about _extras being reserved, etc

File: universum/configuration_support.py (hunk 30 above)
Comment (o-andrieiev) [reply universum/configuration_support.py:180]: should it really be skipped btw?

Comment (i-keliukh) [reply universum/configuration_support.py:182]: The answer is in universum core logic.
The behavior of critical is different depending on where it is set to True.
If it is set in parent, but not in children, all children will be executed, but if any of them fails, the entire execution will stop after that. If critical is set on parent and children, the children will no be executed as soon as one of them fails.

Because of that, in contrast to all other attributes the propagation from parent to child does not happen.

**Review by i-keliukh — APPROVED** [review]

**Review by i-keliukh — DISMISSED** [review]

File: universum/configuration_support.py [hunk 33]
```diff
@@ -1,18 +1,265 @@
...
+
+        >>> cfg = ProjectConfiguration(name='foo', my_var='bar')
+        >>> cfg['name']
+        'foo'
+        >>> cfg['my_var']
+        'bar'
+        >>> cfg['test']
+        """
+        #  _extras are checked first - just in case _extras field is added manually
+        # do note that __setitem__ checks pre_defined fields first, however it's impossible to shadow them by
```
Comment (k-dovgan) [inline universum/configuration_support.py:138]: predefined fields

**Review by i-keliukh — APPROVED** [review]

File: universum/configuration_support.py [hunk 34]
```diff
@@ -1,18 +1,265 @@
...
+        self.if_env_set: str = if_env_set
+        self.children: Optional['Variations'] = None
+        self._extras: Dict[str, str] = {}
+        for key, value in kwargs.items():
+            self._extras[key] = value
+
+    def __repr__(self) -> str:
+        """
+        This function simulates dict-like representation for string output. This is useful when printing contents of
+        :class:`.Variations` objects, as internally they wrap a list of :class:`.ProjectConfiguration` objects
```
Comment (k-dovgan) [inline universum/configuration_support.py:74]: As I understand it, :class:`Variations` without dot in the beginning should work as well, since this is the very same module. And it kinda looks better in my opinion.

**Review by i-keliukh — APPROVED** [review]

**Review by k-dovgan — APPROVED** [review]
